---
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

# Using BLAST to find AMPs from AMP query database on proteomes

```{r, echo = FALSE}
library(tidyverse)
library(ampir)
```

## Proteomes

The proteomes were downloaded from [UniProt](https://www.uniprot.org/proteomes) on 10 June 2021.

**Table 1:** Proteome information from organisms used as subjects 

| Organism Name | Reference proteome ID | Total proteins | Reviewed | Unreviewed |
| --------------- | --------------- | --------------- | --------------- | --------------- |
| *Mus musculus* | [UP000000589](https://www.uniprot.org/proteomes/UP000000589) | 55,470 | 17,068 | 38,402 |
| *Homo sapiens* | | | | | |
| *Rattus norvegicus* | | | | | |
|*Bos taurus* | | | | | |
| | | | | | |

```{r}
mouse_proteome <- read_faa("data/proteomes/M_musculus-proteome-UP000000589.fasta.gz")
```

## BLAST searches

The [BLAST+](https://pubmed.ncbi.nlm.nih.gov/20003500/) version used was blast 2.11.0, build Nov 17 2020 for MacOS.

Each proteome was used to make a local BLAST database using `makeblastdb`. This proteome database was then used to query the AMP dataset with `blastp`, Protein-Protein BLAST 2.11.0+
```{bash, eval = FALSE}
gunzip -dc data/proteomes/M_musculus-proteome-UP000000589.fasta.gz | makeblastdb -in - -title M_musculus-proteome-UP000000589 -dbtype prot -out cache/M_musculus-proteome-UP000000589.fasta
```


```{bash, eval = FALSE}
blastp -db cache/M_musculus-proteome-UP000000589.fasta -query cache/Mus_musculus.fasta -outfmt 6 -max_target_seqs 5 -evalue=10 > data/blastp_results/Mus_musculus.blastp
```

*also do it the other way around, make mus_musculus.fasta the database and query it against M_musculus proteome

            
1. qaccver - Query accession.version (AMPs list ID)
2. saccver - Subject accession.version (Reference proteome ID)
3. pident - Percentage of identical matches
4. length - Alignment length (sequence overlap)
5. mismatch - Number of mismatches
6. gapopen - Number of gap openings
7. qstart - Start of alignment in query
8. qend - End of alignment in query
9. sstart - Start of alignment in subject
10. send - End of alignment in subject
11. evalue - Expect value (the smaller the evalue, the better the homology match)
12. bitscore - Bit score (the higher the bitscore, the better the sequence similarity)

        

```{r}
blast_colnames <- c("qaccver","saccver","pident","length","mismatch","gapopen","qstart","qend","sstart","send","evalue","bitscore")

mouse_amps_blast <- read_tsv("data/blastp_results/Mus_musculus.blastp", col_names = blast_colnames)

mouse_amps_blast %>% group_by(qaccver) %>% slice_max(n = 1, order_by = bitscore)

```


```{r}
mouse_model <- readRDS("models/Mus_musculus_model.rds")

```

```{r, eval=FALSE}
mouse_pred <- predict_amps(mouse_proteome, n_cores = 2, model = mouse_model)
```


```{r}
mouse_pred <- mouse_pred %>% separate(seq_name, into = c("ID", "Description"), sep = "\\s", extra = "merge")
```

